### PR TITLE
Merge pinvoke guidelines into interop guidelines

### LINKED ISSUE #304: organize markdown files
We have lots of markdown files relating to building, debugging, issue management, style, API reviews and suchlike. These are mostly inherited from the original repos and there is duplication, obsolete material, and incorrect material.

Someone needs to take a look and decide what should happen to each doc, how they should be laid out and enumerate what work is needed so several of us can help do it.

@maryamariyan has volunteered to do this.

### PR DESCRIPTION
**Pull request by danmoseley**
Fixes #304

I separated out some edits and the merge.

### CODE REVIEW

File: docs/coding-guidelines/interop-guidelines.md [hunk 1]
```diff
@@ -204,3 +188,273 @@ Guidelines for shim C++ API:
...
+3. `ToString()` allocates yet another managed array **{4}**
+
+That is *{4}* allocations to get a string out of native code. The best you can do to limit this is to reuse the `StringBuilder`
+in another call but this still only saves *1* allocation. It is much better to use and cache a character buffer from `ArrayPool`- you can then get down to just the allocation for the `ToString()` on subsequent calls. 
+
+The other issue with `StringBuilder` is that it always copies the return buffer back up to the first null. If the passed back string isn't terminated or is a double-null-terminated string your P/Invoke is incorrect at best.
+
+If you *do* use `StringBuilder` one last gotcha is that the capacity does **not** include a hidden null which is always accounted for in interop. It is pretty common for people to get this wrong as most APIs want the size of the buffer *including* the null. This can result in wasted/unnecessary allocations.
+
+**[USE]** Char arrays from `ArrayPool` or `StringBuffer` or `stackalloc`
```
Comment (danmosemsft) [inline docs/coding-guidelines/interop-guidelines.md:251]: I added stackalloc here. Probably, we should explicitly talk about buffer allocation in general (not just char buffers)

File: docs/coding-guidelines/interop-guidelines.md [hunk 2]
```diff
@@ -204,3 +188,273 @@ Guidelines for shim C++ API:
...
+
+[1]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.aspx
+[2]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.preservesig.aspx
+[3]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.setlasterror.aspx
+[4]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.exactspelling.aspx
+[5]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.charset.aspx
+
+Strings
+-------
+
```
Comment (jkotas) [inline docs/coding-guidelines/interop-guidelines.md:227]: All/most of these guidelines were folded into the official docs. For example, this section on string is exactly same as  https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#string-parameters (module a few copy edits).

Should we just link to the official docs, and only have dotnet/runtime add-ons (if there are any)?

Comment (danmosemsft) [reply docs/coding-guidelines/interop-guidelines.md:227]: Ha - I hadn't realized how much was the same. @jkoritzinsky looks like you did this in https://github.com/dotnet/docs/pull/9839. Is it obvious to you which parts are specific to us and should be retained?

I suggest we still merge this as a net improvement.

Comment (jkotas) [reply docs/coding-guidelines/interop-guidelines.md:227]: All useful content from `interop-pinvokes.md` was moved over. The differences between the two documents that I can see are:

- The doc on docs.microsoft.com has a few extra details, e.g. about fixed buffers
- The doc on docs.microsoft.com has references to internal implementation details (e.g. names of methods in the C++ runtime impl) deleted.

Comment (danmosemsft) [reply docs/coding-guidelines/interop-guidelines.md:227]: Alright. Shall I just simply delete interop-pinvokes.md and just remove link from interop-guidelines.md ?

Comment (AaronRobinsonMSFT) [reply docs/coding-guidelines/interop-guidelines.md:227]: >Shall I just simply delete interop-pinvokes.md and just remove link from interop-guidelines.md ?

I think so. Our internal guidelines should almost always be the official "best-practices".

Comment (AaronRobinsonMSFT) [reply docs/coding-guidelines/interop-guidelines.md:227]: You could link to the best practices url?

Comment (danmosemsft) [reply docs/coding-guidelines/interop-guidelines.md]: OK, done.

File: docs/coding-guidelines/interop-guidelines.md [hunk 3]
```diff
@@ -156,28 +156,11 @@ internal static partial class Interop
...
-- Use `ArrayPool` for buffer pooling.
-- Be careful of return string termination when allocating buffers (add room for null where needed).
-- Only use `bool` for 32 bit types (matches `BOOL` not `BOOLEAN`).
-- Use `[In]` and `[Out]` only when they differ from the implicit behavior.
-- Explicitly specify the `CharSet` as `Ansi` or `Unicode` when the signature has a string.
-- Use `ExactSpelling` to avoid probing for A/W signature variants.
-- Do not set `PreserveSig` to false.
+When defining the P/Invoke signatures and structs, we follow the guidelines in the [interop best practices documentation](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices).
 
+ Interop signatures / structs / constants should be defined using the same name / capitalization / etc. that's used in the corresponding native code. We should not rename any of these based on managed coding guidelines. The only exception to this is for the constant grouping type, which should be named with the most discoverable name possible; if that name is a concept (e.g. Errors), it can be named using managed naming guidelines.
```
Comment (danmosemsft) [inline docs/coding-guidelines/interop-guidelines.md:163]: I left this bit in as the doc only recommends constants be named this way.

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: docs/coding-guidelines/interop-guidelines.md (hunk 3 above)
Comment (danmosemsft) [reply docs/coding-guidelines/interop-guidelines.md:163]: BTW that topic probably should mention stackalloc given how often we use it for interop buffers. And suggest a max size.

Comment (jkotas) [reply docs/coding-guidelines/interop-guidelines.md:163]: The doc has the recommendation for everything. It just has the text about the constant grouping type replaced with "CONSIDER" recommendation (vs. DO recommendation for method names) .

> DO use the same naming and capitalization for your methods and parameters as the native method you want to call.
> CONSIDER using the same naming and capitalization for constant values.

Comment (jkotas) [reply docs/coding-guidelines/interop-guidelines.md:163]: There is a whole chapter on the constant grouping types before this one.

Comment (danmosemsft) [reply docs/coding-guidelines/interop-guidelines.md:163]: Right - we were saying "should" vs the doc just says CONSIDER. I am happy to remove it though.

Comment (danmosemsft) [reply docs/coding-guidelines/interop-guidelines.md]: removed it

**Review by jkotas — APPROVED** [review]

**Pull request merged**